Mathlib/Computability/PrimrecFilter.lean

```diff
@@ -0,0 +1,209 @@
+/-
+Copyright (c) 2025 David J. Webb. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David J. Webb
+-/
+import Mathlib.Computability.Primrec
+/-!
+
+# Bounded quantifiers are primitive recursive
+This file contains lemmata for showing bounded existentially and universally quantified
+statements are primitive recursive, as well as more general filtering for arbitrary
+primcodable types.
+
+## Main results
+- Filtering for elements from a list that meet a primrec criterion is primrec
+- Checking whether a list has some element meeting a primrec criterion is primrec
+- Checking whether every element of a list meets a primrec criterion is primrec
+- Primitive recursive functions are closed under bounded existential and universal quantifiers
+
+## References
+- [R. I. Soare *Turing Computability - Theory and Applications*] [Soare2016]
+-/
+
+open List
+open Primrec
+
+namespace Primrec
+
+/-- Filtering a list for elements that satisfy a decidable predicate is primitive recursive. -/
+lemma list_filter {α} [Primcodable α] (f : α → Prop) [DecidablePred f]
+(hf : PrimrecPred f) : Primrec fun L => (filter (fun a => f a) L) := by
+  rw [← List.filterMap_eq_filter]
+  apply listFilterMap Primrec.id
+  simp [Primrec₂, Option.guard]
+  exact ite (PrimrecPred.comp hf snd) (option_some_iff.mpr snd) (const none)
+
+/-- Checking if any element of a list satisfies a decidable predicate is primitive recursive. -/
+lemma filter_exists {α} [Primcodable α] (f : α → Prop) [DecidablePred f]
+    (hf : PrimrecPred f) : PrimrecPred fun (L : List α) => (∃ a ∈ L, f a) := by
+  let g := fun L => List.filter (fun a => f a) L
+  have h (L : List α): ((g L).length ≠ 0) ↔ (∃ a ∈ L, f a) := by simp [g]
+  apply PrimrecPred.of_eq ?_ h
+  apply PrimrecPred.not
+  apply PrimrecRel.comp Primrec.eq ?_ (const 0)
+  exact comp list_length (list_filter f hf)
+
+/-- Checking if every element of a list satisfies a decidable predicate is primitive recursive. -/
+lemma filter_forall {α} [Primcodable α] (f : α → Prop) [DecidablePred f]
+    (hf : PrimrecPred f) : PrimrecPred fun (L : List α) => (∀ a ∈ L, f a) := by
+  let g := fun L => List.filter (fun a => f a) L
+  have h (L : List α): ((g L).length = L.length) ↔ (∀ a ∈ L, f a) := by simp [g]
+  apply PrimrecPred.of_eq ?_ h
+  refine PrimrecRel.comp Primrec.eq ?_ list_length
+  exact comp list_length (list_filter f hf)
+
+/-- Bounded existential quantifiers are primitive recursive. -/
+lemma bounded_exists (f : ℕ → Prop) [DecidablePred f] (hf : PrimrecPred f) :
+    PrimrecPred fun n => ∃ x < n, f x := by
+  have h : PrimrecPred fun n => (∃ a ∈ (range n), f a) := by
+    apply PrimrecPred.comp (filter_exists f hf) list_range
```

Comment:
Can you make this `apply` an `exact`?
Once you have done so, can you remove the `by exact`? That's slightly shorter.
(Also in the next lemma.)